### PR TITLE
[JENKINS-67173] [JEP-235] Remove Callable allowlist and FilePath agent-to-controller support

### PR DESCRIPTION
**Pull request by daniel-beck**
This PR does the following:

* It removes the ability to disable the agent-to-controller security subsystem.
* It removes the `Callable` allowlist, whose default entries are all long obsolete and everyone else had 7 years to adapt their plugins.
* It removes the custom file access rules that allow `FilePath` methods limited access to the controller file system.

The following sections discuss why they're no longer needed or justified.

Additionally, there's a [JEP submission](https://github.com/jenkinsci/jep/pull/381) for this change, which gets continually updated with new information from telemetry.

## Agent-to-controller security subsystem general toggle

It's existed since 2014. It's the default for new installs since 2016. The option exists in case of unforeseen problems when it was introduced, so I'm pretty sure we can assume that there are no significant problems.

Additionally, there are only very narrow circumstances in which it is a good idea to disable this (probably as narrow as not having CSRF protection, which is always enabled as well).

## Callable Allowlist

An extensible allowlist for specific `Callable`s to be executed. This only applies to those built with dependencies pre-dating the 2014 introduction of the `RoleSensitive` interface. All plugins in the [default allowlist](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/jenkins/security/s2m/callable.conf) have long since been updated.

> hudson.maven.MavenBuildProxy$Filter$AsyncInvoker

https://github.com/jenkinsci/maven-plugin/commit/3a4b06f1fd8e317af2926bab6be137feb19e7895 (released Oct 2014: https://github.com/jenkinsci/maven-plugin/releases/tag/maven-plugin-2.7.1)

> com.cloudbees.plugins.deployer.engines.Engine$FingerprintDecorator

Obsolete: https://github.com/jenkins-infra/update-center2/blob/953eae236debefb7f9ed0777e935c6cb12f0d632/resources/artifact-ignores.properties#L28

> hudson.scm.SubversionWorkspaceSelector$1

https://github.com/jenkinsci/subversion-plugin/commit/1a2d547100d3fc391c152dad54c4235a86838552 (released June 2016: https://github.com/jenkinsci/subversion-plugin/releases/tag/2.6.0)

> org.jenkinsci.plugins.gitclient.CliGitAPIImpl$GetPrivateKeys

https://github.com/jenkinsci/git-client-plugin/pull/147 (released Oct 2014: https://github.com/jenkinsci/git-client-plugin/releases/tag/git-client-1.11.0)

> com.cloudbees.jenkins.plugins.sshcredentials.SSHAuthenticator$1

https://github.com/jenkinsci/ssh-credentials-plugin/commit/45e1d5e3a9103a4d48d47407aedabd82b198667a (released May 2016: https://github.com/jenkinsci/ssh-credentials-plugin/releases/tag/ssh-credentials-1.12)

> com.synopsys.arc.jenkinsci.plugins.cygwinprocesskiller.CygwinProcessKiller$KillerRemoteCall

https://github.com/jenkinsci/cygwin-process-killer-plugin/commit/1851e8092e0f1e971e252bf5d08db4588d16e2ab (released Jan 2018: https://github.com/jenkinsci/cygwin-process-killer-plugin/releases/tag/cygwin-process-killer-0.2)

> hudson.plugins.selenium.JenkinsCapabilityMatcher$LabelMatcherCallable

https://github.com/jenkinsci/selenium-plugin/commit/0b77252fc88ba9ac3ab2a7faf7b5a3a4da61bbc1 (released April 2016: https://github.com/jenkinsci/selenium-plugin/releases/tag/selenium-2.53.0; but the plugin is an unsafe, unmaintained mess anyway)

----

Besides the default allowlist, everyone has now had 7 years to implement `#checkRoles`. Plugins need a core dependency older than 1.580.1 or 1.587 to even compile.

While plugins are being distributed that implement `Callable` without `RoleSensitive`, only two ( https://plugins.jenkins.io/docker-build-step/ and https://plugins.jenkins.io/terraform/ ) have been released in the last five years, the former is the only one with more than 5k installs, and there's a good chance they all don't even send their `Callable`s in the agent-to-controller direction.

## Custom File Access Rules

This covers methods in `FilePath` that can be called from agent processes and are allowed to access _some_ directories in the controller based on a customizable allowlist.

As [today's security advisory](https://www.jenkins.io/security/advisory/2021-11-04/) demonstrates, this is perhaps not the best idea. While we've addressed all known issues, the mere presence of so many vulnerabilities is an indication we should scrap it.

To the best of my knowledge, this is not used a lot, since people need to write pretty weird code to begin with.

The uses we know about (see https://github.com/daniel-beck/jep/tree/jep-submission-a2c-security/jep/0000#incompatible-plugins) can be (and in parts already have been) adapted to not use this feature.

Given various limitations (problems with custom Jenkins home layout -- long known -- and symlinks -- new today), this should just be scrapped.

People who _really_ need this should write `SlaveToMasterCallable`s ([not `SlaveToMasterFileCallable`](https://github.com/jenkinsci/jenkins/pull/5884)) and carefully review https://www.jenkins.io/doc/developer/security/remoting-callables/ before attempting this.

# Compatibility

In terms of the behavior change:

* There should be few `Callable`s / `FileCallable`s that do not implement `RoleSensitive` left, and even fewer that need to be sent in the agent-to-controller direction.
* `FilePath` callers are more difficult to identify statically, but #5890 adds telemetry that identifies any such calls. https://github.com/jenkinsci/jep/pull/381 tracks a list of known affected plugins.

Regarding deleted types and methods: I checked `usage-in-plugins` with the following file, and only found `configuration-as-code` (toggling `AdminWhitelistRule` which is retained _enough_) and some proprietary CloudBees plugins.

```
# General
jenkins/security/s2m/AdminWhitelistRule
jenkins/security/s2m/ConfigDirectory
jenkins/security/s2m/ConfigFile
jenkins/security/s2m/MasterKillSwitchConfiguration
jenkins/security/s2m/MasterKillSwitchWarning
# FilePathFilter
jenkins/security/s2m/AdminFilePathFilter
jenkins/security/s2m/DefaultFilePathFilter
jenkins/security/s2m/FilePathRuleConfig
jenkins/security/s2m/FilePathRule
jenkins/security/s2m/OpMatcher
jenkins/security/s2m/RunningBuildFilePathFilter
jenkins/ReflectiveFilePathFilter
jenkins/SoloFilePathFilter
jenkins/ReflectiveFilePathFilter
jenkins/FilePathFilterAggregator
jenkins/FilePathFilter
# Callables
jenkins/security/s2m/AdminCallableMonitor
jenkins/security/s2m/AdminCallableWhitelist
jenkins/security/s2m/CallableDirectionChecker
jenkins/security/s2m/CallableRejectionConfig
jenkins/security/s2m/CallableWhitelist
jenkins/security/s2m/CallableWhitelistConfig
jenkins/security/s2m/RejectedCallable
```

### Proposed changelog entries

* The agent-to-controller security subsystem is now always enabled. The admin-customizable allowlists for callables and file paths have been removed. The ability to access some files on the controller from agents has been removed. Some plugins are incompatible with this change and may need to be updated, see [the issue tracker](https://issues.jenkins.io/issues/?jql=%22Epic%20Link%22%20%3D%20JENKINS-67173). ([JEP-235](https://github.com/jenkinsci/jep/tree/master/jep/235))

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/FilePath.java [hunk 1]
```diff
@@ -579,7 +577,7 @@ public Void invoke(File dir, VirtualChannel channel) throws IOException, Interru
         }
         private static final long serialVersionUID = 1L;
     }
-    private static class UnzipLocal extends SecureFileCallable<Void> {
+    private static class UnzipLocal extends MasterToSlaveFileCallable<Void> {
```
Comment (jglick) [inline core/src/main/java/hudson/FilePath.java:580]: merits a comment explaining this

**Review by jglick — APPROVED** [review]

**Comment by jglick** [comment]
Follows up #5880, FTR.

File: core/src/main/java/hudson/FilePath.java (hunk 1 above)
Comment (daniel-beck) [reply core/src/main/java/hudson/FilePath.java:580]: Per https://github.com/jenkinsci/jenkins/commit/104c751d907919dd53f5090f84d53c671a66457b it actually _is_ transferred: A controller can `untar`/`unzip` a file on an agent to a directory on the agent.

Comment (jglick) [reply core/src/main/java/hudson/FilePath.java:580]: Right, just needs a source comment saying that.

File: core/src/main/java/hudson/FilePath.java [hunk 2]
```diff
@@ -2849,9 +2833,9 @@ private static void readFromTar(String name, File baseDir, InputStream in) throw
                     if (parent != null) mkdirs(parent);
```
Comment (jglick) [inline core/src/main/java/hudson/FilePath.java:2833]: We ought to be able to delete the `getCanonicalFile` calls here as well (merge conflict with #5886).

File: test/src/test/java/jenkins/security/AgentToControllerSecurityTest.java [hunk 3]
```diff
@@ -107,6 +109,43 @@ public Void call() throws Exception {
         }
     }
 
+    // ----- Agent to controller access control can be disabled using system property (but really shouldn't)
+    @Test
+    public void ensureBypass() throws Exception {
+        CallableDirectionChecker.BYPASS = true;
```
Comment (jglick) [inline test/src/test/java/jenkins/security/AgentToControllerSecurityTest.java:115]: or `FlagRule`

**Review by jglick — APPROVED** [review]

**Comment by jglick** [comment]
merge conflicts (modify/delete I guess)

**Comment by jglick** [comment]
Also needs to delete `core/src/main/java/jenkins/telemetry/impl/SlaveToMasterFileCallableUsage.java`.

**Review by jglick — APPROVED** [review]
:fire:

**Review by jvz — APPROVED** [review]
Nice to see this removed!

**Review by timja — APPROVED** [review]

**Review by basil — APPROVED** [review]
> “If we wish to count lines of code, we should not regard them as ‘lines produced’ but as ‘lines spent.’ ” — Edsger W. Dijkstra

By the same logic, we should regard excellent PRs like this not as ‘lines deleted’ but as ‘lines earned.’

**Comment by daniel-beck** [comment]
A quarter of instances (~73k) are now reporting agent-to-controller `FilePath` access telemetry (which I believe to be the most notable backward compatibility issue here), see https://github.com/jenkinsci/jep/pull/381. Of those, only around 0.75% report any actual accesses, most in `log-parser` and `cobertura`, all of which are fixed in latest releases or have a PR.

46 instances report uses in [Publish Over SSH](https://plugins.jenkins.io/publish-over-ssh/), a plugin with no maintainers. I checked and [the code path](https://issues.jenkins.io/browse/JENKINS-67236) involves [Infrastructure for Publish Over X](https://plugins.jenkins.io/publish-over/) (another plugin without maintainers) so probably not trivial to resolve. The affected feature is the "SSH key path" configuration option which causes the agent to read an SSH key from a file from the controller (in an allowed path). A workaround seems to be to provide the SSH key inline, with one limitation: it looks like environment variables can be involved in the file path resolution, so some jobs may be configured to have dynamic SSH key lookup. With few reported uses and a workaround for (likely) most cases, this doesn't justify considering this plugin to be a blocker to merging this PR. I'd rather merge this now than wait for more information and have a shorter cycle to fix towards the next LTS baseline.

Thoughts?

**Review by oleg-nenashev — APPROVED** [review]
I am +1 for that, it is a good security hardening for a potentially dangerous use-case. And thanks for not adding `unmaintained mess anyway` to the Cygwin Process Killer plugin...

I guess CloudBees Jenkins Operations Center implementations were verified, right?

**Comment by oleg-nenashev** [comment]
FWIW it would be nice to fix Publish Over SSH before this patch is released

**Comment by oleg-nenashev** [comment]
We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Comment by basil** [comment]
```
67 files changed, 336 insertions(+), 4226 deletions(-)
```

:fire:

**Comment by basil** [comment]
With this PR, `org.jenkinsci.plugins.workflow.support.steps.ExecutorStepTest#accessPermittedOnlyFromCurrentBuild` fails. Without this PR, the same test passes.

**Comment by jglick** [comment]
I will deal with that.

**Comment by timja** [comment]
Redirect doesn't exist for https://www.jenkins.io/redirect/AdminWhitelistRule

It's spamming everyones logs with JCasc as well, probably better to just remove full stop?

**Comment by daniel-beck** [comment]
> It's spamming everyones logs with JCasc as well, probably better to just remove full stop?

No preference; it's specifically left in for JCasC to not break that. I would expect log spam to be easier to deal with that it breaking entirely on update?

**Pull request merged**